app/controllers/protocols_controller.rb

```diff
@@ -586,17 +582,185 @@ def import
+    # id 17= expected result
+    whitelist_complex = %w(8 9 15 18 19 20)
+    # (complex mapping with nested hashes) id 8 = software package,
+    # id 9 = dataset, id 15 = command, id 18 = attached sub protocol
+    # id 19= safety information ,
+    # id 20= regents (materials, like scinote samples kind of)
+    @json_object['steps'].each do |step| # loop over steps
+      pos += 1 # position of step (first, second.... etc),
+      @db_json['steps'][pos.to_s] = {} # the json we will insert into db
+      @db_json['steps'][pos.to_s]['position'] = pos
+      step['components'].each do |key, value|
+        # sometimes there are random index values as keys
+        # instead of hashes, this is a workaround to that buggy json format
+        key = value if value.class == Hash
+        if whitelist_simple.include?(key['component_type_id'])
+          # append is the string that we append values into for description
+          case key['component_type_id']
+          when '1'
+            @db_json = protocolsio_step_description_populate(@db_json, key, pos)
+          when '6'
+            @db_json = protocolsio_step_title_populate(@db_json, key, pos)
+          when '17'
+            @db_json = protocolsio_step_expected_result_populate(
+              @db_json, key, pos
+            )
+          end
+          # (complex mapping with nested hashes)
+          # id 8 = software package, id 9 = dataset,
+          # id 15 = command, id 18 = attached sub protocol
+          # id 19= safety information ,
+          # id 20= regents (materials, like scinote samples kind of)
+        elsif key && whitelist_complex.include?(key['component_type_id'])
+          case key['component_type_id']
+          when '8'
+            @db_json = protocolsio_step_software_package_populate(
+              @db_json, key, pos
+            )
+          when '9'
+            @db_json = protocolsio_step_dataset_populate(@db_json, key, pos)
+          when '15'
+            @db_json = protocolsio_step_command_populate(@db_json, key, pos)
+          when '18'
+            @db_json = protocolsio_step_attached_sub_protocol_populate(
+              @db_json, key, pos
+            )
+          when '19'
+            @db_json = protocolsio_step_safety_information_populate(
+              @db_json, key, pos
+            )
+          end
+        end # finished step component case iteration
+      end # finished looping over step components
+    end # steps
+    protocol = nil
+    respond_to do |format|
+      transaction_error = false
+      @protocolsio_general_error = false
+      Protocol.transaction do
+        begin
+          protocol = import_new_protocol(
+            @db_json, current_team, params[:type].to_sym, current_user
+          )
+        rescue Exception
+          transaction_error = true
+          raise ActiveRecord:: Rollback
+        end
+      end
+      p_name =
+        if @db_json['name'].present? && !@db_json['name'].empty?
```

Comment:
I think the second part (!@db_json['name'].empty?) is redundant.
Just use: @db_json['name'].present?